dashboard/src/actions/index.js

```diff
@@ -197,3 +198,196 @@ export const addScratchpadEntry = entry => ({
 export const deleteScratchpadEntries = () => ({
     type: "DELETE_SCRATCHPAD_ENTRIES"
 });
+
+export const updateShareId = shareId => ({
+    type: "UPDATE_SHARE_ID",
+    shareId
+});
+
+export const queryFound = found => ({
+    type: "QUERY_FOUND",
+    found: found
+});
+
+const doShareMutation = (dispatch, getState) => {
+    let query = getState().query.text,
+        stringifiedQuery = JSON.stringify(encodeURI(query)),
+        mutation = `
+    mutation {
+        set {
+            <_:share> <_share_> ${stringifiedQuery} .
+        }
+    }`;
+
+    fetch(dgraphQuery(false), {
+        method: "POST",
+        mode: "cors",
+        headers: {
+            Accept: "application/json",
+            "Content-Type": "text/plain"
+        },
+        body: mutation
+    })
+        .then(checkStatus)
+        .then(response => response.json())
+        .then(function handleResponse(result) {
+            if (result.uids && result.uids.share) {
+                dispatch(updateShareId(result.uids.share));
+            }
+        })
+        .catch(function(error) {
+            console.log(error.stack);
+            return error.message;
+        })
+        .then(function(errorMsg) {
+            if (errorMsg !== undefined) {
+                dispatch(
+                    saveErrorResponse(
+                        "Got error while saving querying for share: " + errorMsg
+                    )
+                );
+            }
+        });
+};
+
+export const getShareId = (dispatch, getState) => {
+    let query = getState().query.text;
+    if (query === "") {
+        return;
+    }
+    let stringifiedQuery = JSON.stringify(encodeURI(query)),
+        // Considering that index is already set on the pred, schema mutation
+        // should be a no-op. Lets see if we have already stored this query by
+        // performing an exact match.
+        checkQuery = `
+mutation {
+    schema {
+        _share_: string @index(exact) .
+    }
+}
+{
+    query(func:eq(_share_, ${stringifiedQuery})) {
+        _uid_
+    }
+}`;
+
+    timeout(
+        6000,
```

Comment:
Any reason that we need to use `setTimetout`? It would be great if we can do without it.